src/PluginStructs/P200_data_struct.cpp

```diff
@@ -0,0 +1,189 @@
+  reset();
+}
+
+void P200_data_struct::reset() {
+  if (easySerial != nullptr) {
+    delete easySerial;
+    easySerial = nullptr;
+  }
+}
+
+bool P200_data_struct::init(ESPEasySerialPort port, const int16_t serial_rx, const int16_t serial_tx, unsigned long baudrate) {
+  if ((serial_rx < 0) && (serial_tx < 0)) {
+    return false;
+  }
+  reset();
+  easySerial = new (std::nothrow) ESPeasySerial(port, serial_rx, serial_tx);
+
+  if (isInitialized()) {
+    easySerial->begin(baudrate);
+    return true;
+  }
+  return false;
+}
+
+bool P200_data_struct::isInitialized() const {
+  return easySerial != nullptr;
+}
+
+bool P200_data_struct::loop() {
+  if (!isInitialized()) {
+    return false;
+  }
+  bool fullSentenceReceived = false;
+
+  if (easySerial != nullptr) {
+    int available = easySerial->available();
+
+    while (available > 0 && !fullSentenceReceived) {
+      char c = easySerial->read();
+      --available;
+
+      if (available == 0) {
+        available = easySerial->available();
+        if (available == 0) { resetSerialTimeout(); }
+        delay(0);
+      }
+
+      switch (sentence_part.length()) {
+        case 0:
+          if (c != '\r') break; // wait for '\r'
+        default:
+          sentence_part += c;
+          sentence_checksum -= c;
+          break;
+      }
+
+      if (max_length_reached()) { fullSentenceReceived = true; }
+    }
+  }
+
+  if (serialTimeout() || fullSentenceReceived) {
+    const size_t length = sentence_part.length();
+    bool valid          = length > 0;
+
+    fullSentenceReceived = false;
+
+    if (valid) {
+    // message or full buffer
+      
+      // RAW
+      if (output_type == 2) {
+        length_last_received = sentence_part.length();
+        last_sentence = base64::encode(sentence_part);
+        fullSentenceReceived = true;
+        sentence_part = "";
+        ++sentences_received;
+      } else {
+        // Cheksum
+        sentence_checksum = sentence_checksum & 0xff;
+
+        if (sentence_checksum != 0) {
+          sentence_checksum = 0;
+          sentence_part = "";
+          ++sentences_received_error;
+        } else {
+          // JSON & CSV
+          String field = "";
+          uint16_t field_start = 2;
+          uint16_t field_end = sentence_part.indexOf('\r', field_start);
+          uint16_t field_separator = 0;
+          if (output_type == 0) { last_sentence = '{'; }
+
+          while (field_end <= length) {
+            // this will drop last field (checksum) 
+            if (last_sentence.length() >= 2) { last_sentence += ','; }
+            field = sentence_part.substring(field_start, field_end); // <Field-Labe><Tab><Field-Value>
+
+            field_separator = field.indexOf('\t');
+            if (output_type == 0) { last_sentence += '\"'; }
+            last_sentence += field.substring(0, field_separator); // <Field-Label>
+            if (output_type == 0) { last_sentence += '\"'; }
+            last_sentence += ':';
```

Comment:
There are json support functions available to correctly output json elements, these should better be used (check `src/src/Helpers/StringConverter.h&.cpp`

Reply:
Changed code to use StringConverter helper.